### ISSUE TITLE
[chore] Make some internal packages publicly available for importing

### ISSUE THREAD

**Issue by odubajDT** [open-telemetry/opentelemetry-collector-contrib#36615]
### Component(s)

internal/k8stest
internal/core

### Describe the issue you're reporting

There is a existing use-case, where multiple internal functions are used in other non-opentelemetry collector repositories to test custom built collectors which are based on the OTel-collector. To avoid the need to copy over the functions needed to properly test the custom collectors, it would be beneficial to have those functions available publicly outside of the `internal` package, since in the current state, Golang does not allow importing packages marked as internal.

I would propose to make the following packages publicly available:

 - [k8stest](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/internal/k8stest)
 - [idutils](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/internal/coreinternal/idutils)
 
 My suggestion would be the following:
 
 - `internal/k8stest` -> `pkg/k8stest`
 - `internal/coreinternal/idutils` -> `pkg/core/idutils`
 
 Looking forward to your feedback!

**Comment by atoulme** [comment]
That is very broad, would you please help identify which functions you aim to make reusable?

**Comment by odubajDT** [comment]
> That is very broad, would you please help identify which functions you aim to make reusable?

Basically, we would need most of the functions (if not all) from `k8stest` package and functions from [idutils](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/internal/coreinternal/idutils) package. Already adapted the issue description.

Would this be possible?

**Comment by mx-psi** [comment]
What is the use case here?

**Comment by odubajDT** [comment]
We're building a custom version of the collector with a subset of components. 

With this collector we're doing an additional e2e testing for our custom configurations, where in these tests we are using functions from `k8stest` and `idutils` packages, which are in the current state copied to the collector repository. It would help us if these packages will be publicly available for importing to avoid copying them.

**Comment by braydonk** [comment]
Triage: Discussion is underway, removing `needs-triage`.

**Comment by TylerHelmuth** [comment]
These packages change enough that I don't really want to worry about our public footprint or breaking changes. Is the functionality specific to something else we do in the collector? Instead of being public can you duplicate them?

**Comment by odubajDT** [comment]
> These packages change enough that I don't really want to worry about our public footprint or breaking changes. Is the functionality specific to something else we do in the collector?

I don't see any breaking changes, nor any functionality changes in the `idutils` package according to the [git history](https://github.com/open-telemetry/opentelemetry-collector-contrib/commits/main/internal/coreinternal/idutils) since August 2021. I would consider this pretty stable from this reason.

Regarding the `k8stest` package, this one exists (as a separate package) since June 2023 ([this](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/e794019c8038b4d52ba469a2ce5f0735ac5a44b5#diff-336d27a0d59204be713137d2998f73a6a56db6da1d7f803ec5aba1649453d895) commit). Outside of maintenance work (dependency bumps, lint adaptations etc,) I found only the following two commits adapting the logic (mostly due to bug fixes and supporting newer version of k8s) [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/d8f1fe70a1a63acf9a2ca802f410d3098729cd0e#diff-69c122cfe9873a36807d97e549dbb5f7da2ed53ffece6283b2cbbc83253666d5) and [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/ee145c5b2b92aac8fcf1a96ed6a2c79ae9380726#diff-336d27a0d59204be713137d2998f73a6a56db6da1d7f803ec5aba1649453d895)

For me as a potential user of the library (when it will be piblicly available) this is not an issue.

> Instead of being public can you duplicate them?

In the current state we have them duplicated. We would like to get rid of the duplications.

**Comment by atoulme** [comment]
`idutils` is mostly used in conjunction with zipkin and jaeger, and a bit in a load generator. It is probably going to be fine to use, but please add tests and add a metadata.yaml if you expose it, and make yourself codeowner.

No idea about `k8stest`.

**Comment by odubajDT** [comment]
> `idutils` is mostly used in conjunction with zipkin and jaeger, and a bit in a load generator. It is probably going to be fine to use, but please add tests and add a metadata.yaml if you expose it, and make yourself codeowner.
> 
> No idea about `k8stest`.

Can you specify what kind of additional tests are you expecting? Seems like the functions in the package are all well unit tested.

**Comment by evan-bradley** [comment]
>These packages change enough that I don't really want to worry about our public footprint or breaking changes.

@TylerHelmuth I agree with being mindful about our public API, but could we indicate that these packages have no stability guarantees? We do the same with in development components, and that seems to work out okay from what I can tell. I think so long as we're up-front with what level of support these are given, we should be okay.

>Is the functionality specific to something else we do in the collector?

I think these are useful mostly outside the upstream Collector repos, particularly for distro authors like Dynatrace. We use them in our distribution to ensure we know if any changes break our config examples. [Here's an example](https://github.com/Dynatrace/dynatrace-otel-collector/tree/main/internal/testbed/integration/k8senrichment). I can see other distro authors, either vendors or non-vendors, using them to help ensure they have automated tests in place to help them understand the impact of updating the versions of the components in their distro.

>Instead of being public can you duplicate them?

That's what we've done for now for the sake of getting something out the door, but we would prefer to be able to automatically pull in updates from upstream and streamline the process of making contributions if we need changes that would also make sense for the community. Handling breaking changes isn't an issue from our end.

**Comment by odubajDT** [comment]
@TylerHelmuth your thoughts?

**Comment by odubajDT** [comment]
> We don't have a precedent for packages with no stability guarantees at all. I am okay with doing that but I wonder if they should be in a separate folder/module set/have a different versioning scheme (like v0.0.1)?

I don't have anything against it, but why a separate versioning is needed? Since the current version still starts with `v0`, I don't see it as a problem.

**Comment by mx-psi** [comment]
@odubajDT Even for v0 modules we have some restrictions in place which are documented [here](https://github.com/open-telemetry/opentelemetry-collector/blob/main/docs/coding-guidelines.md#breaking-changes). This is mentioned explicitly [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/pkg#pkg-folder) for the `pkg` folder.

**Comment by odubajDT** [comment]
> @odubajDT Even for v0 modules we have some restrictions in place which are documented [here](https://github.com/open-telemetry/opentelemetry-collector/blob/main/docs/coding-guidelines.md#breaking-changes). This is mentioned explicitly [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/pkg#pkg-folder) for the `pkg` folder.

Ok, I see thank you. But regarding the maintenance burden and the stability, I did some investigation and summed up the findings in [this comment](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/36615#issuecomment-2519490814). 

Would it still be an issue with stability? We can see that these packages are quite stable.

**Comment by odubajDT** [comment]
Any news on this? @mx-psi @TylerHelmuth @evan-bradley

**Comment by TylerHelmuth** [comment]
I am willing to unblock if other maintainers are.

**Comment by mx-psi** [comment]
I am willing to unblock this as well. We could use the `x` prefix that we have adopted in the core repository to signal the lack of stability guarantees if needed.

**Comment by evan-bradley** [comment]
Item 6 here contains the guidelines for the prefix: https://github.com/open-telemetry/opentelemetry-collector/blob/main/docs/coding-guidelines.md#module-organization

It doesn't apply perfectly here. I would be okay either adding an addendum to this rule that describes adding an experimental module for which there is no existing stable module, or just making it abundantly clear in the module readmes that these are experimental.

**Comment by odubajDT** [comment]
PRs got merged, thank you!
